File: java/TokenEncoder.java
public class TokenEncoder {

  // CREDITS: http://vodi.de for https://github.com/SmartStandards/Snowflake44

  /**
   * Decodes a long to a token string (of maximum 12 characters).
   * The String will be PascalCase (e.g. "HelloWorld")
  */
  public static String decode(long int64Value) throws Exception {
    String rawToken = TokenEncoder.int64ToRaw(int64Value);
    String token = TokenEncoder.rawToReadable(rawToken);
    return token;
  }

  public static long encode(String token) throws Exception {
    String rawToken = TokenEncoder.readableToRaw(token);
    return TokenEncoder.rawToInt64(rawToken);
  }

  /**
   * Decodes a long to a raw token string (of maximum 12 characters).
   * The String will be CAPITALS and underscores (e.g. "HELLO_WORLD")
  */
  public static String int64ToRaw(long int64Value) throws Exception {
    long bitmask = 31;
    int offset = 0;
    byte mappedCode;
    byte unmappedCode;
    char[]decoded = new char[12];

    if (int64Value < 0) throw new Exception("Negative values like " + int64Value + " are not supported!");

    for (int i = 0; i <= 11; i++) {
      mappedCode = (byte)((int64Value & bitmask) >> offset);

      if (mappedCode >= 1 && mappedCode <= 26) {
        unmappedCode = (byte)(mappedCode | 64);
      } else if (mappedCode == 27) {
        unmappedCode = (byte)196; // Ä (Windows-1252 aka ANSI)
      } else if (mappedCode == 28) {
        unmappedCode = (byte)214; // Ö (Windows-1252 aka ANSI)
      } else if (mappedCode == 29) {
        unmappedCode = (byte)220; // Ü (Windows-1252 aka ANSI)
      } else if (mappedCode == 30) {
        unmappedCode = (byte)223; // ß (Windows-1252 aka ANSI)
      } else if (mappedCode == 31) {
        unmappedCode = (byte)35; // #
      } else {
        unmappedCode = (byte)32; // Space
      }
      decoded[i] = (char)(unmappedCode & 0xFF); // & 0xFF ist the conversion from signed byte to unsigned
      bitmask <<= 5;
      offset += 5;
    }

    return new String(decoded).stripTrailing().replace(' ', '_');
  }

  public static long rawToInt64(String rawToken) throws Exception { 

    if (rawToken.length() > 12) {
      throw new Exception("Maximum length of 12 characters was exceeded by \"" + rawToken + "\"!");
    }

    if (rawToken.endsWith("_")) {
      throw new Exception("String must not end with underscore: \"" + rawToken + "\"!");
    }

    String toEncode = TokenEncoder.toUpperCaseFixed(rawToken);

    if (toEncode.length() < 12) toEncode = TokenEncoder.padRight(toEncode,12, '_');

    long encodedValue = 0;
    int offset = 0;
    long mappedCode;

    for (var i = 0; i <= 11; i++) {

      mappedCode = TokenEncoder.getMappedCode(toEncode.charAt(i));

      if (mappedCode > 31) {
        throw new Exception("Unsupported Character '" + toEncode.charAt(i) + "' in \"" + rawToken + "\"!");
      }
      encodedValue |= (mappedCode << offset);
      offset += 5;
    }
    return encodedValue;
  }

  public static String readableToRaw(String token) throws Exception {

    if ((token) == null) return null;

      if ((token == "")) return "";

      char digitSurrogate = ' ';
      char escapedChar = ' ';
      boolean withinEscapeSequence = false;
      boolean upperCaseExpected = true;

      StringBuilder sb = new StringBuilder(token.length() * 2);

      for (int i = 0; i <= (token.length() - 1); i++) {

        char c = token.charAt(i);

        digitSurrogate = ' ';
        escapedChar = ' ';

        if (c == '1') { digitSurrogate = 'I'; }
        if (c == '2') { digitSurrogate = 'Z'; }
        if (c == '3') { digitSurrogate = 'E'; }
        if (c == '4') { digitSurrogate = 'H'; }
        if (c == '5') { digitSurrogate = 'S'; }
        if (c == '6') { digitSurrogate = 'G'; }
        if (c == '7') { digitSurrogate = 'L'; }
        if (c == '8') { digitSurrogate = 'B'; }
        if (c == '9') { digitSurrogate = 'P'; }
        if (c == '0') { digitSurrogate = 'O'; }

        if (c == '#') { escapedChar = '#'; }
        if (c == '+') { escapedChar = 'A'; }
        if (c == ',') { escapedChar = 'C'; }
        if (c == '.') { escapedChar = 'D'; }
        if (c == '?') { escapedChar = 'F'; }
        if (c == '(') { escapedChar = 'J'; }
        if (c == ')') { escapedChar = 'K'; }
        if (c == '-') { escapedChar = 'M'; }
        if (c == '&') { escapedChar = 'N'; }
        if (c == '@') { escapedChar = 'Q'; }
        if (c == ':') { escapedChar = 'R'; }
        if (c == '\'') { escapedChar = 'T'; }
        if (c == '\\') { escapedChar = 'U'; }
        if (c == '/') { escapedChar = 'V'; }
        if (c == '=') { escapedChar = 'W'; }
        if (c == '*') { escapedChar = 'X'; }
        if (c == '!') { escapedChar = 'Y'; }
        if (c == ']') { escapedChar = 'Ä'; }
        if (c == '[') { escapedChar = 'Ö'; }
        if (c == '%') { escapedChar = 'Ü'; }
        if (c == '^') { escapedChar = 'ß'; }

        // Terminate escaping state

        if (withinEscapeSequence && digitSurrogate == ' ' && (escapedChar == ' ' || escapedChar == '_')) {
          withinEscapeSequence = false;
          sb.append('_');
          upperCaseExpected = true;
        }

        if (digitSurrogate != ' ') {
          if (!withinEscapeSequence) sb.append('#');
          withinEscapeSequence = true;
          sb.append(digitSurrogate);
          continue;
        }

        if (escapedChar != ' ') {
          if (!withinEscapeSequence) {
            sb.append('#');
            upperCaseExpected = true;
          }
          sb.append(escapedChar);
          continue;
        }

        char upperC = Character.toUpperCase(c);

        long mappedCode = TokenEncoder.getMappedCode(upperC);

        if (mappedCode > 31) {
          throw new Exception("Unsupported Character '" + c + "' in \"" + token + "\"!");
        }

        if (Character.isUpperCase(c) != upperCaseExpected) sb.append('_');

        sb.append(upperC);

        upperCaseExpected = false;
      } // next

      return sb.toString();
  }

  /**
   * Transforms a string containing separate words into a PascalCase formatted string.
   * E.g. "HELLO_WORLD" becomes "HelloWorld"
  */
  public static String rawToReadable(String rawToken) {

    if (rawToken == null) return null;

      if (rawToken == "") return "";

      StringBuilder sb = new StringBuilder(rawToken.length() * 2);

      boolean nextOneUp = true; // Der Erste immer Groß
      boolean escaping = false;
      boolean withinDigits = false;

      for (int i = 0; i <= (rawToken.length() - 1); i++) {

        char c = rawToken.charAt(i);

        if (withinDigits) {
          if (c == '_') {
            withinDigits = false;
            escaping = false;
            nextOneUp = true;
            continue;
          }
        }

        if (escaping) {
          if (c == 'I') { sb.append('1'); withinDigits = true; continue; }
          if (c == 'Z') { sb.append('2'); withinDigits = true; continue; }
          if (c == 'E') { sb.append('3'); withinDigits = true; continue; }
          if (c == 'H') { sb.append('4'); withinDigits = true; continue; }
          if (c == 'S') { sb.append('5'); withinDigits = true; continue; }
          if (c == 'G') { sb.append('6'); withinDigits = true; continue; }
          if (c == 'L') { sb.append('7'); withinDigits = true; continue; }
          if (c == 'B') { sb.append('8'); withinDigits = true; continue; }
          if (c == 'P') { sb.append('9'); withinDigits = true; continue; }
          if (c == 'O') { sb.append('0'); withinDigits = true; continue; }

          if (c == '#') { sb.append('#'); }
          if (c == 'A') { sb.append('+'); }
          if (c == 'C') { sb.append(','); }
          if (c == 'D') { sb.append('.'); }
          if (c == 'F') { sb.append('?'); }
          if (c == 'J') { sb.append('('); }
          if (c == 'K') { sb.append(')'); }
          if (c == 'M') { sb.append('-'); }
          if (c == 'N') { sb.append('&'); }
          if (c == 'Q') { sb.append('@'); }
          if (c == 'R') { sb.append(':'); }
          if (c == 'T') { sb.append('\''); }
          if (c == 'U') { sb.append('\\'); }
          if (c == 'V') { sb.append('/'); }
          if (c == 'W') { sb.append('='); }
          if (c == 'X') { sb.append('*'); }
          if (c == 'Y') { sb.append('!'); }
          if (c == 'Ä') { sb.append(']'); }
          if (c == 'Ö') { sb.append('['); }
          if (c == 'Ü') { sb.append('%'); }
          if (c == 'ß') { sb.append('^'); }

          if (!withinDigits) {
            escaping = false;
            nextOneUp = true;
          }
          continue;
        }

        if (c == '#') {
          escaping = true;
          continue;
        }

        if (c == '_') {
          nextOneUp = !nextOneUp;
          continue;
        }

        if (nextOneUp) {
          sb.append(Character.toUpperCase(c));
        } else {
          sb.append(Character.toLowerCase(c));
        }

        nextOneUp = false;
      }
      return sb.toString();
  }

  public static long getMappedCode(char c) {

    byte unmappedCode = (byte)c; // byte casting uses ISO 8859-1 (aka Windows-1252) mapping

    // int unsignedRepresentation = unmappedCode & 0xff;

    if (unmappedCode == 95) { // _
      return 0;
    } else if (65 <= unmappedCode && unmappedCode <= 90) { // A-Z
      return (long)(unmappedCode - 64);
    } else if (unmappedCode == 35) { // #
      return 31;
    } else if (unmappedCode == -60) { // Ä,ä (unsigned representation 196)
      return 27;
    } else if (unmappedCode == -42) { // Ö,ö (unsigned representation 214)
      return 28;
    } else if (unmappedCode == -36 ) { // Ü,ü (unsigned representation 220)
      return 29;
    } else if (unmappedCode == -33 ) { // ß (unsigned representation 223)
      return 30;
    } else {
      // System.out.println(c + ": " +  unmappedCode + " " + unsignedRepresentation);
      return 32;
    }
  }

  public static String padRight(String input, int length, char paddingChar) {
    if (input.length() >= length) {
        return input;
    }
    StringBuilder sb = new StringBuilder(input);
    while (sb.length() < length) {
        sb.append(paddingChar);
    }
    return sb.toString();
  }

  public static String toUpperCaseFixed(String incoming){
    char[] chars = incoming.toCharArray();
    for (int i = 0; i < chars.length; i++) {
      chars[i] = Character.toUpperCase(chars[i]);
    }
    return new String(chars);
  } 

}

File: js/lib/TokenEncoder.js
// ENCODING OF THIS FILE IS NOT UTF-8, BUT ISO 8859-1 (aka Windows-1252)!

class TokenEncoder {

  static decode(int64Value) {
    var rawToken = TokenEncoder.int64ToRaw(int64Value);
    var decodedToken = TokenEncoder.rawToReadable(rawToken);
    return decodedToken;
  }

  static encode(token){
    var rawToken = TokenEncoder.readableToRaw(token);
    return TokenEncoder.rawToInt64(rawToken);
  }

  static rawToInt64(rawToken) {

    if (rawToken.length > 12) {
      throw new Error('Maximum length of 12 characters was exceeded by "' + rawToken + '"!');
    }

    if (rawToken.endsWith("_")) {
      throw new Error('String must not end with underscore: "' + rawToken + '"!');
    }

    var toEncode = TokenEncoder.toUpperCaseFixed(rawToken);

    if (toEncode.length < 12) toEncode = toEncode.padEnd(12, '_');

    var encodedValue = 0n; // the n makes it BigInt
    var offset = 0n;
    var mappedCode = 0n;

    for (var i = 0; i <= 11; i++) {

      mappedCode = BigInt(TokenEncoder.getMappedCode(toEncode, i));

      if (mappedCode > 31) {
        throw new Error('Unsupported Character "' + toEncode[i] + '" in "' + rawToken + '"!');
      }
      encodedValue |= (mappedCode << offset);
      offset += 5n;
    }
    return encodedValue;
  }

  static int64ToRaw(int64Value) {

    if ((int64Value < 0)) int64Value *= -1n;

    var bitmask = 31n; // the n makes it BigInt
    var offset = 0n;
    var mappedCode = 0n;
    var unmappedCode = 0n;
    var decodedChars = [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '];
    var decodedString = "";

    for (var i = 0; i <= 11; i++) {
      mappedCode = ((int64Value & bitmask) >> offset);

      if (mappedCode >= 1 && mappedCode <= 26) {
        unmappedCode = (mappedCode + 64n);
      } else if (mappedCode == 27) {
        unmappedCode = 196; // ƒ (ISO 8859-1 aka Windows-1252)
      } else if (mappedCode == 28) {
        unmappedCode = 214; // ÷ (ISO 8859-1 aka Windows-1252)
      } else if (mappedCode == 29) {
        unmappedCode = 220; // ‹ (ISO 8859-1 aka Windows-1252)
      } else if (mappedCode == 30) {
        unmappedCode = 223; // ﬂ (ISO 8859-1 aka Windows-1252)
      } else if (mappedCode == 31) {
        unmappedCode = 35; // #
      } else { 
        unmappedCode = 32; // Space
      }

      //decodedString += String.fromCharCode(Number(unmappedCode)); // concatenation is faster than array push!
      decodedChars[i] = String.fromCharCode(Number(unmappedCode)); // hopefully this is even faster

      bitmask <<= 5n;
      offset += 5n;
    }

    //decodedString = decodedString.trimEnd().replaceAll(' ','_');
    decodedString = decodedChars.join('').trimEnd().replaceAll(' ','_');

    return decodedString;
  }

  static readableToRaw(token) {

    if ((token) == null) return null;

    if ((token == "")) return "";

    var digitSurrogate = ' ';
    var escapedChar = ' ';
    var withinEscapeSequence = false;
    var upperCaseExpected = true;

    var sb = "";

    for (var i = 0; i < (token.length); i++) {

      var c = token[i];

      digitSurrogate = ' ';
      escapedChar = ' ';

      if (c == '1') { digitSurrogate = 'I'; }
      if (c == '2') { digitSurrogate = 'Z'; }
      if (c == '3') { digitSurrogate = 'E'; }
      if (c == '4') { digitSurrogate = 'H'; }
      if (c == '5') { digitSurrogate = 'S'; }
      if (c == '6') { digitSurrogate = 'G'; }
      if (c == '7') { digitSurrogate = 'L'; }
      if (c == '8') { digitSurrogate = 'B'; }
      if (c == '9') { digitSurrogate = 'P'; }
      if (c == '0') { digitSurrogate = 'O'; }

      if (c == '#') { escapedChar = '#'; }
      if (c == '+') { escapedChar = 'A'; }
      if (c == ',') { escapedChar = 'C'; }
      if (c == '.') { escapedChar = 'D'; }
      if (c == '?') { escapedChar = 'F'; }
      if (c == '(') { escapedChar = 'J'; }
      if (c == ')') { escapedChar = 'K'; }
      if (c == '-') { escapedChar = 'M'; }
      if (c == '&') { escapedChar = 'N'; }
      if (c == '@') { escapedChar = 'Q'; }
      if (c == ':') { escapedChar = 'R'; }
      if (c == '\'') { escapedChar = 'T'; }
      if (c == '\\') { escapedChar = 'U'; }
      if (c == '/') { escapedChar = 'V'; }
      if (c == '=') { escapedChar = 'W'; }
      if (c == '*') { escapedChar = 'X'; }
      if (c == '!') { escapedChar = 'Y'; }
      if (c == ']') { escapedChar = 'ƒ'; }
      if (c == '[') { escapedChar = '÷'; }
      if (c == '%') { escapedChar = '‹'; }
      if (c == '^') { escapedChar = 'ﬂ'; }

      // Terminate escaping state

      if (withinEscapeSequence && digitSurrogate == ' ' && (escapedChar == ' ' || escapedChar == '_')) {
        withinEscapeSequence = false;
        sb += '_';
        upperCaseExpected = true;
      }

      if (digitSurrogate != ' ') {
        if (!withinEscapeSequence) sb += '#';
        withinEscapeSequence = true;
        sb += digitSurrogate;
        continue;
      }

      if (escapedChar != ' ') {
        if (!withinEscapeSequence) {
          sb += '#';
          upperCaseExpected = true;
        }
        sb += escapedChar;
        continue;
      }

      var upperC = TokenEncoder.toUpperCaseFixed(c);

      var mappedCode = BigInt(TokenEncoder.getMappedCode(upperC,0));

      if (mappedCode > 31) {
        throw new Error('Unsupported Character "' + c + '" in "' + token + '"!');
      }

      var isAlreadyUpper = (c == upperC);

      if (isAlreadyUpper != upperCaseExpected) sb += '_';

      sb += upperC;

      upperCaseExpected = false;
    } // next

    return sb;
  }

  static rawToReadable(rawToken) {

    if (rawToken == null) return null;

    if (rawToken == "") return "";

    var pascalCaseString = "";

    var nextOneUp = true; // Der Erste immer Groﬂ
    var escaping = false;
    var withinDigits = false;

    for (var i = 0; i <= (rawToken.length - 1); i++) {

      var c = rawToken[i];

      if (withinDigits) {
        if (c == '_') {
          withinDigits = false;
          escaping = false;
          nextOneUp = true;
          continue;
        }
      }

      if (escaping) {
        if (c == 'I') { pascalCaseString += '1'; withinDigits = true; continue; }
        if (c == 'Z') { pascalCaseString += '2'; withinDigits = true; continue; }
        if (c == 'E') { pascalCaseString += '3'; withinDigits = true; continue; }
        if (c == 'H') { pascalCaseString += '4'; withinDigits = true; continue; }
        if (c == 'S') { pascalCaseString += '5'; withinDigits = true; continue; }
        if (c == 'G') { pascalCaseString += '6'; withinDigits = true; continue; }
        if (c == 'L') { pascalCaseString += '7'; withinDigits = true; continue; }
        if (c == 'B') { pascalCaseString += '8'; withinDigits = true; continue; }
        if (c == 'P') { pascalCaseString += '9'; withinDigits = true; continue; }
        if (c == 'O') { pascalCaseString += '0'; withinDigits = true; continue; }

        if (c == '#') { pascalCaseString += '#'; }
        if (c == 'A') { pascalCaseString += '+'; }
        if (c == 'C') { pascalCaseString += ','; }
        if (c == 'D') { pascalCaseString += '.'; }
        if (c == 'F') { pascalCaseString += '?'; }
        if (c == 'J') { pascalCaseString += '('; }
        if (c == 'K') { pascalCaseString += ')'; }
        if (c == 'M') { pascalCaseString += '-'; }
        if (c == 'N') { pascalCaseString += '&'; }
        if (c == 'Q') { pascalCaseString += '@'; }
        if (c == 'R') { pascalCaseString += ':'; }
        if (c == 'T') { pascalCaseString += '\''; }
        if (c == 'U') { pascalCaseString += '\\'; }
        if (c == 'V') { pascalCaseString += '/'; }
        if (c == 'W') { pascalCaseString += '='; }
        if (c == 'X') { pascalCaseString += '*'; }
        if (c == 'Y') { pascalCaseString += '!'; }
        if (c == 'ƒ') { pascalCaseString += ']'; }
        if (c == '÷') { pascalCaseString += '['; }
        if (c == '‹') { pascalCaseString += '%'; }
        if (c == 'ﬂ') { pascalCaseString += '^'; }

        if (!withinDigits) {
          escaping = false;
          nextOneUp = true;
        }
        continue;
      }

      if (c == '#') {
        escaping = true;
        continue;
      }

      if (c == '_') {
        nextOneUp = !nextOneUp;
        continue;
      }

      if (nextOneUp) {
        c = TokenEncoder.toUpperCaseFixed(c);
      } else {
        c = c.toLowerCase();
      }
      pascalCaseString += c;
      nextOneUp = false;
    }
    return pascalCaseString;
  }

  static toUpperCaseFixed(incoming){
    var outgoing = "";
    for (var i = 0; i <= (incoming.length - 1); i++) {
      var c = incoming[i];
      if (c != 'ﬂ') c = c.toUpperCase(); // JavaScript thinks that "ﬂ" should be capitalized to "SS" - we don't want this.
      outgoing += c;
    }
    return outgoing;
  }

  static getMappedCode(rawToken, i) {

    var unmappedCode = rawToken.charCodeAt(i);

    if (unmappedCode == 95) { // _
      return 0n;
    } else if (65 <= unmappedCode && unmappedCode <= 90) { // A-Z
      return (unmappedCode - 64);
    } else if (unmappedCode == 35) { // #
      return 31;
    } else if (unmappedCode == 196) { // ƒ,‰
      return 27;
    } else if (unmappedCode == 214) { // ÷,ˆ
      return 28;
    } else if (unmappedCode == 220) { // ‹,¸
      return 29;
    } else if (unmappedCode == 223) { // ﬂ
      return 30;
    } else {
      return 32;
    }
  }
}